src/v/cluster/topics_frontend.cc

```diff
@@ -949,4 +982,137 @@ topics_frontend::do_cancel_moving_partition_replicas(
     co_return results;
 }
 
+ss::future<std::error_code> topics_frontend::change_replication_factor(
+  model::topic_namespace topic,
+  cluster::replication_factor_type new_replication_factor,
+  model::timeout_clock::time_point timeout) {
+    auto tp_metadata = _topics.local().get_topic_metadata_ref(topic);
+    if (!tp_metadata.has_value()) {
+        co_return errc::topic_not_exists;
+    }
+    if (!tp_metadata.value().get().is_topic_replicable()) {
+        co_return errc::topic_operation_error;
+    }
+
+    auto current_replication_factor
+      = tp_metadata.value().get().get_replication_factor();
+
+    if (current_replication_factor == new_replication_factor) {
+        co_return errc::success;
+    }
+
+    new_replicas_assigment new_replicas;
+
+    // We should hold allocation_units from reallocate_partition while we do
+    // replicate_and_wait.
+    std::optional<allocation_units> allocation_units;
+
+    if (new_replication_factor < current_replication_factor) {
+        auto res = co_await decrease_replication_factor(
+          topic, new_replication_factor);
+
+        if (!res.has_value()) {
+            co_return errc::invalid_request;
+        }
+        new_replicas = res.value();
+    } else {
+        auto res = co_await increase_replication_factor(
+          topic, new_replication_factor, allocation_units);
+
+        if (!res.has_value()) {
+            co_return errc::invalid_request;
+        }
+        new_replicas = res.value();
+    }
+
+    move_topic_replicas_cmd cmd(topic, std::move(new_replicas));
+
+    co_return co_await replicate_and_wait(
+      _stm, _features, _as, std::move(cmd), timeout);
+}
+
+ss::future<result<topics_frontend::new_replicas_assigment>>
+topics_frontend::increase_replication_factor(
+  model::topic_namespace topic,
+  cluster::replication_factor_type new_replication_factor,
+  std::optional<allocation_units>& units) {
+    std::vector<move_topic_replicas_data> cmds;
+
+    auto tp_metadata = _topics.local().get_topic_metadata(topic);
+    if (!tp_metadata.has_value()) {
+        co_return errc::topic_not_exists;
+    }
+
+    for (const auto& assignment : tp_metadata->get_assignments()) {
+        if (assignment.replicas.size() > new_replication_factor) {
+            co_return errc::topic_invalid_replication_factor;
+        }
+
+        auto partition_constraints = cluster::partition_constraints(
+          assignment.id, new_replication_factor);
+        auto stable_assigments = partition_assignment(
+          assignment.group, assignment.id, assignment.replicas);
+
+        auto ntp = model::ntp(topic.ns, topic.tp, assignment.id);
+
+        auto reallocation = co_await _allocator.invoke_on(
+          partition_allocator::shard,
+          [partition_constraints = std::move(partition_constraints),
+           stable_assigments = std::move(stable_assigments),
+           ntp](partition_allocator& al) {
+              return al.reallocate_partition(
```

Comment:
Now that we're making it easy to bulk-create many new partitions by changing a topic's replication factor, I think it is necessary to extend `reallocate_partition` to call `check_cluster_limits` in some form: otherwise the user can exceed the cluster-wide limits on memory per partition etc.

This can be in a separate PR but we should do it before release.

Reply:
It will be added later with health_monitor. Like in partition balancer